servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/KeepAliveManager.java

```diff
@@ -0,0 +1,315 @@
+     *     <li>{@link Future} instance to timeout ack of PING sent.</li>
+     *     <li>{@link #KEEP_ALIVE_ACK_TIMEDOUT} if we fail to receive a PING ack for the configured timeout.</li>
+     *     <li>{@link #CLOSED} if the channel is closed.</li>
+     * </ul>
+     */
+    @Nullable
+    private Object keepAliveState;
+    @Nullable
+    private final GenericFutureListener<Future<? super Void>> pingWriteCompletionListener;
+
+    KeepAliveManager(final Channel channel, @Nullable final KeepAlivePolicy keepAlivePolicy) {
+        this(channel, keepAlivePolicy, (task, delay, unit) ->
+                channel.eventLoop().schedule(task, delay, unit),
+                (ch, idlenessThresholdSeconds, onIdle) -> ch.pipeline().addLast(
+                        new IdleStateHandler(idlenessThresholdSeconds, idlenessThresholdSeconds, 0) {
+                            @Override
+                            protected void channelIdle(final ChannelHandlerContext ctx, final IdleStateEvent evt) {
+                                onIdle.run();
+                            }
+                        }));
+    }
+
+    KeepAliveManager(final Channel channel, @Nullable final KeepAlivePolicy keepAlivePolicy,
+                     final Scheduler scheduler, final IdlenessDetector idlenessDetector) {
+        this.channel = channel;
+        this.scheduler = scheduler;
+        if (keepAlivePolicy != null) {
+            disallowKeepAliveWithoutActiveStreams = !keepAlivePolicy.withoutActiveStreams();
+            pingAckTimeoutNanos = keepAlivePolicy.ackTimeout().toNanos();
+            pingWriteCompletionListener = future -> {
+                if (future.isSuccess() && keepAliveState == KEEP_ALIVE_ACK_PENDING) {
+                    // Schedule a task to verify ping ack within the pingAckTimeoutMillis
+                    keepAliveState = scheduler.afterDuration(() -> {
+                        if (keepAliveState != null) {
+                            keepAliveState = KEEP_ALIVE_ACK_TIMEDOUT;
+                            LOGGER.debug(
+                                    "channel={}, timeout {}ns waiting for keep-alive PING(ACK), writing go_away.",
+                                    this.channel, pingAckTimeoutNanos);
+                            channel.writeAndFlush(new DefaultHttp2GoAwayFrame(NO_ERROR))
+                                    .addListener(f -> {
+                                        if (f.isSuccess()) {
+                                            LOGGER.debug("Closing channel={}, after keep-alive timeout.", this.channel);
+                                            KeepAliveManager.this.close0();
+                                        }
+                                    });
+                        }
+                    }, pingAckTimeoutNanos, NANOSECONDS);
+                }
+            };
+            int idleInSeconds = (int) keepAlivePolicy.idleDuration().getSeconds();
```

Comment:
KeepAlivePolicyBuilder applies a limit when setting the Idle duration, although I think it is safer to avoid overflow here and then we don't have to worry about applying a limit upfront or missing applying the limit if someone implement their own `KeepAlivePolicy` etc..